src/pudl/output/ferc1.py

```diff
@@ -1677,3 +1790,508 @@ def remove_intra_table_calculated_values(exploded: pd.DataFrame) -> pd.DataFrame
+                # TODO: need to check that there are no conflicts between tags that are
+                # being propagated, e.g. if two different ancestors have been tagged
+                # rate_base: yes and rate_base: no.
+                leaf_tags |= pruned_forest.nodes[node]["tags"]
+                # Root nodes have no weight because they don't come from calculations
+                # We assign them a weight of 1.0
+                if not pruned_forest.nodes[node].get("weight", False):
+                    assert node in roots
+                    node_weight = 1.0
+                else:
+                    node_weight = pruned_forest.nodes[node]["weight"]
+                leaf_weight *= node_weight
+
+            # Construct a dictionary describing the leaf node and convert it into a
+            # single row DataFrame. This makes adding arbitrary tags easy.
+            leaf_attrs = {
+                "xbrl_factoid_original": pruned_forest.nodes[leaf][
+                    "xbrl_factoid_original"
+                ],
+                "weight": leaf_weight,
+                "tags": leaf_tags,
+                "source_table": leaf.source_table,
+                "xbrl_factoid": leaf.xbrl_factoid,
+            }
+            leaf_rows.append(pd.json_normalize(leaf_attrs, sep="_"))
+
+        # Combine the two dataframes we've constructed above:
+        return (
+            pd.merge(leafy_meta, pd.concat(leaf_rows), validate="one_to_one")
+            .convert_dtypes()
+            .set_index(["source_table", "xbrl_factoid"])
+        )
+
+    @property
+    def root_calculations(self: Self) -> pd.DataFrame:
+        """Produce an exploded metadataframe containing only roots and leaves.
+
+        This dataframe has a format similar to exploded_meta and can be used in
+        conjunction with the exploded data to verify that the root values can still
+        be correctly calculated from the leaf values.
+
+        """
+
+        def leafy_meta_to_calculations(df: pd.DataFrame) -> str:
+            return json.dumps(
+                [
+                    {
+                        "name": row.xbrl_factoid,
+                        "weight": float(row.weight),
+                        "xbrl_factoid_original": row.xbrl_factoid_original,
+                        "source_tables": [row.source_table],
+                    }
+                    for row in df.itertuples()
+                ]
+            )
+
+        root_calcs: pd.DataFrame = (
+            self.leafy_meta.reset_index()
+            .groupby(["root_table", "root_xbrl_factoid"], as_index=False)
+            .apply(leafy_meta_to_calculations)
+        )
+        root_calcs.columns = ["root_table", "root_xbrl_factoid", "calculations"]
+        return root_calcs
+
+    @property
+    def table_names(self: Self) -> list[str]:
+        """Produce the list of tables involved in this explosion."""
+        return list(self.exploded_meta.reset_index()["table_name"].unique())
```

Comment:
I think this is correct, but it's also kind of duplicative of information in the `Exploder` that exists outside of the tree context.